Mathlib/Algebra/SkewMonoidAlgebra/Basic.lean

```diff
@@ -0,0 +1,1481 @@
+  classical
+  simp_rw [mul_apply, sum]
+  rw [Finsupp.sum_comm, Finsupp.sum_single_index]
+  · apply Finset.sum_eq_zero
+    simp_rw [ite_eq_right_iff]
+    rintro g'' _ rfl
+    exact absurd ⟨_, rfl⟩ h
+  · simp_rw [Finsupp.sum, smul_zero, mul_zero, ite_self, Finset.sum_const_zero]
+
+theorem single_mul_apply_aux [Monoid G] [MulSemiringAction G k] (f : SkewMonoidAlgebra k G) {r : k}
+    {x y z : G} (H : ∀ a, x * a = y ↔ a = z) : (single x r * f) y = r * x • f z := by
+  classical
+  have : (f.sum fun a b => ite (x * a = y) (0 * x • b) 0) = 0 := by
+    simp only [zero_mul, ite_self,sum_zero]
+  calc
+    (HMul.hMul (α := SkewMonoidAlgebra k G) (single x r) f) y =
+        sum f fun a b => ite (x * a = y) (r * x • b) 0 :=
+      (mul_apply _ _ _).trans <| sum_single_index this
+    _ = f.sum fun a b => ite (a = z) (r * x • b) 0 := by simp only [H]
+    _ = if z ∈ f.support then r * x • f z else 0 := (f.support.sum_ite_eq' _ _)
+    _ = _ := by split_ifs with h <;> simp [support] at h <;> simp [h]
+
+theorem single_one_mul_apply [Monoid G] [MulSemiringAction G k] (f : SkewMonoidAlgebra k G) (r : k)
+    (x : G) : (single (1 : G) r * f) x = r * f x := by
+  rw [single_mul_apply_aux, one_smul]
+  simp only [one_mul, forall_const]
+
+theorem single_mul_apply_of_not_exists_mul [Monoid G] [MulSemiringAction G k] (r : k) {g g' : G}
+    (x : SkewMonoidAlgebra k G) (h : ¬∃ d, g' = g * d) : (single g r * x) g' = 0 := by
+  classical
+  rw [mul_apply, sum_single_index]
+  · apply Finset.sum_eq_zero
+    simp_rw [ite_eq_right_iff]
+    rintro g'' _hg'' rfl
+    exact absurd ⟨_, rfl⟩ h
+  · simp_rw [sum, Finsupp.sum, zero_mul, ite_self, Finset.sum_const_zero]
+
+end MiscTheorems
+
+section AddHom
+
+variable [AddCommMonoid k]
+
+theorem addHom_ext {N : Type*} [AddZeroClass N] ⦃f g : (SkewMonoidAlgebra k G) →+ N⦄
+    (H : ∀ x y, f (single x y) = g (single x y)) : f = g := by
+  have hf (x : G →₀ k) {f : (SkewMonoidAlgebra k G) →+ N} :
+    f (ofFinsupp x) = (AddMonoidHom.comp f
+      ⟨⟨ofFinsupp, ofFinsupp_zero⟩, (fun _ _ ↦ ofFinsupp_add)⟩) x := by rfl
+  simp_rw [hf] at H
+  convert Finsupp.addHom_ext H
+  refine' ⟨fun heq => by rw [heq], fun heq => by ext x; rw [← eta x]; simp_rw [hf, heq]⟩
+
+/-- `single` as an `AddMonoidHom`.
+
+See `lsingle` for the stronger version as a linear map. -/
+@[simps]
+def singleAddHom (a : G) : k →+ SkewMonoidAlgebra k G where
+  toFun := single a
+  map_zero' := single_zero a
+  map_add' _ := single_add a _
+
+lemma singleAddHom_apply (a : G) (b : k) : ((singleAddHom a) b) = single a b := by
+  simp only [singleAddHom, AddMonoidHom.coe_mk, ZeroHom.coe_mk]
+
+@[ext high]
+theorem addHom_ext' {N : Type*} [AddZeroClass N] ⦃f g : (SkewMonoidAlgebra k G) →+ N⦄
+    (H : ∀ x, f.comp (singleAddHom x) = g.comp (singleAddHom x)) : f = g :=
+  addHom_ext fun x => DFunLike.congr_fun (H x)
+
+end AddHom
+
+/-! #### Non-unital, non-associative algebra structure -/
+section NonUnitalNonAssocAlgebra
```

Comment:
```suggestion

section NonUnitalNonAssocAlgebra
```

Reply:
Is the suggested change to delete the comment? We took it from the structure of MonoidAlgebra.

Reply:
No, sorry if the diff is unclear. It's adding a blank line between the comment and the section